src/server/verification/processor/provider/ZoomProvider.js

```diff
@@ -31,83 +38,130 @@ class ZoomProvider implements IEnrollmentProvider {
-    const { api, logger } = this
+    const { api, logger, _supportsFeature } = this
     const log = customLogger || logger
+    const { defaultMinimalMatchLevel, defaultSearchIndexName } = api
+    const { LivenessCheckFailed, SecurityCheckFailed, FacemapNotFound } = ZoomAPIError
+
     // send event to onEnrollmentProcessing
     const notifyProcessor = async eventPayload => onEnrollmentProcessing(eventPayload)
 
     // throws custom exception related to the predefined verification cases
     // e.g. livenes wasn't passed, duplicate found etc
     const throwCustomException = (customMessage, customResponse, zoomResponse = {}) => {
       const exception = new Error(customMessage)
+      // removing debug fields
+      let redactedResponse = omit(zoomResponse, 'callData', 'additionalSessionData', 'serverInfo')
+
+      // removing all large data (e.g. images , facemaps)
+      redactedResponse = omitBy(redactedResponse, (_, field) => field.endsWith('Base64'))
 
       exception.response = {
-        // removing all large data (e.g. images , facemaps)
-        ...pick(zoomResponse, 'code', 'subCode', 'message'),
+        ...redactedResponse,
         ...customResponse,
         isVerified: false
       }
 
       throw exception
     }
 
-    // 1. checking for duplicates
-    let duplicate
-    let faceSearchResponse
-    const { defaultMinimalMatchLevel } = api
+    let alreadyEnrolled
+    // checking is API supports delete enrollment endpoint
+    const isDisposeSupported = await _supportsFeature(ZoomAPIFeature.DisposeEnrollment)
 
+    // 1. checking if facescan already uploaded & enrolled
     try {
-      const { results, ...response } = await api.faceSearch(payload, defaultMinimalMatchLevel, customLogger)
+      await api.readEnrollment(enrollmentIdentifier, customLogger)
+      alreadyEnrolled = true
+    } catch (exception) {
+      // if something other that 'FacemapNotFound was thrown - re-throwing
+      if (FacemapNotFound !== exception.name) {
+        throw exception
+      }
 
-      faceSearchResponse = response
-      // excluding own enrollmentIdentifier
-      duplicate = results.find(
-        ({ enrollmentIdentifier: matchId }) => matchId.toLowerCase() !== enrollmentIdentifier.toLowerCase()
-      )
+      // otherwise, setting alreadyEnrolled to false
+      alreadyEnrolled = false
+    }
+
+    // 2. performing liveness check and storing facescan / audit trail images (if need)
+    let isLive = true
+
+    try {
+      const callArgs = [enrollmentIdentifier, payload, customLogger]
+      const enroll = () => api.submitEnrollment(...callArgs)
+
+      // if already enrolled, will call checkLiveness
+      // this is need because enroll doesn't re-checks facemap/images
+      // for liveness if identifier already enrolled
+      if (alreadyEnrolled) {
+        // it doesn't need enrollmentIdentifier param,
+        // so we're slicing it from the start of the args
+        await api.checkLiveness(...slice(callArgs, 1))
```

Comment:
we dont do liveness check, both search and enroll already verify that facemap/scan has liveness

Reply:
1. search requires the facemap will be enrolled and stored to the database. It doesn't works with the raw facemap / facescan as before. It consumes enrollmentIdentifier of the patter to search instead of base64 facemap received from UI
2. Enroll doesn't checks liveness if enrollment already exists. It immediately throws with 'already exists' exception and stops processing

Reply:
alreadyEnrolled flow updated to call match-3d according to suggestions in #290